packages/svelte/src/lib/component/shared.ts

```diff
@@ -0,0 +1,108 @@
+  const fileTypes = config ? Object.keys(config) : [];
+
+  const maxFileCount = config
+    ? Object.values(config).map((v) => v.maxFileCount)
+    : [];
+
+  return { fileTypes, multiple: maxFileCount.some((v) => v && v > 1) };
+};
+
+export const allowedContentTextLabelGenerator = (
+  config?: ExpandedRouteConfig,
+): string => {
+  return capitalizeStart(INTERNAL_doFormatting(config));
+};
+
+export const capitalizeStart = (str: string) => {
+  return str.charAt(0).toUpperCase() + str.slice(1);
+};
+
+const INTERNAL_doFormatting = (config?: ExpandedRouteConfig): string => {
+  if (!config) return "";
+
+  const allowedTypes = Object.keys(config) as (keyof ExpandedRouteConfig)[];
+
+  const formattedTypes = allowedTypes.map((f) => {
+    if (f.includes("/")) return `${f.split("/")[1].toUpperCase()} file`;
+    return f === "blob" ? "file" : f;
+  });
+
+  // Format multi-type uploader label as "Supports videos, images and files";
+  if (formattedTypes.length > 1) {
+    const lastType = formattedTypes.pop();
+    return `${formattedTypes.join("s, ")} and ${lastType}s`;
+  }
+
+  // Single type uploader label
+  const key = allowedTypes[0];
+  const formattedKey = formattedTypes[0];
+
+  const { maxFileSize, maxFileCount } = config[key]!;
+
+  if (maxFileCount && maxFileCount > 1) {
+    return `${formattedKey}s up to ${maxFileSize}, max ${maxFileCount}`;
+  } else {
+    return `${formattedKey} (${maxFileSize})`;
+  }
+};
+
+export type UploadthingComponentProps<TRouter extends FileRouter> = {
+  [TEndpoint in keyof TRouter]: {
+    endpoint: TEndpoint;
+
+    onUploadProgress?: (progress: number) => void;
+    onUploadBegin?: (fileName: string) => void;
+    onClientUploadComplete?: (res?: UploadFileResponse[]) => void;
+    onUploadError?: (error: UploadThingError<inferErrorShape<TRouter>>) => void;
+  } & (undefined extends inferEndpointInput<TRouter[TEndpoint]>
+    ? // eslint-disable-next-line @typescript-eslint/ban-types
+      {}
+    : {
+        input: inferEndpointInput<TRouter[TEndpoint]>;
+      });
+}[keyof TRouter];
+
+export const progressWidths: Record<number, string> = {
+  0: "after:w-0",
+  10: "after:w-[10%]",
+  20: "after:w-[20%]",
+  30: "after:w-[30%]",
+  40: "after:w-[40%]",
+  50: "after:w-[50%]",
+  60: "after:w-[60%]",
+  70: "after:w-[70%]",
+  80: "after:w-[80%]",
+  90: "after:w-[90%]",
+  100: "after:w-[100%]",
+};
+
+/**
+ * @example
+ * const uploader = createUploader<OurFileRouter>({
+ *   endpoint="someEndpoint",
+ *   onUploadComplete={(res) => console.log(res)},
+ *   onUploadError={(err) => console.log(err)},
+ * })
+ */
+export function createUploader<TRouter extends FileRouter>(
+  props: FileRouter extends TRouter
+    ? ErrorMessage<"You forgot to pass the generic">
+    : UploadthingComponentProps<TRouter>,
+) {
+  // Cast back to UploadthingComponentProps<TRouter> to get the correct type
+  // since the ErrorMessage messes it up otherwise
+  return props as UploadthingComponentProps<TRouter>;
```

Comment:
Is it intentional to just return props here?

Reply:
Yes. You cannot just call a Svelte component with a generic type like you can in React, the type must be passed indirectly via props. This function is just a convenience helper to type the uploader props